In [1]:
import numpy as np 
import pandas as pd 

In [2]:
df = pd.read_csv("FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding = "latin-1")

In [3]:
df.head(10)

Area Code     Area  Item Code              Item  Element Code  \
0          4  Algeria       2501        Population           511   
1          4  Algeria       2501        Population          5301   
2          4  Algeria       2901       Grand Total           664   
3          4  Algeria       2901       Grand Total           674   
4          4  Algeria       2901       Grand Total           684   
5          4  Algeria       2903  Vegetal Products           664   
6          4  Algeria       2903  Vegetal Products           674   
7          4  Algeria       2903  Vegetal Products           684   
8          4  Algeria       2941   Animal Products           664   
9          4  Algeria       2941   Animal Products           674   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   
5           Food supply (kcal/capita/day)  kcal/capita/day   2932.00   
6  Protein supply quantity (g/capita/day)     g/capita/day     67.14   
7      Fat supply quantity (g/capita/day)     g/capita/day     53.84   
8           Food supply (kcal/capita/day)  kcal/capita/day    444.00   
9  Protein supply quantity (g/capita/day)     g/capita/day     27.76   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28  
5   2958.00   2941.00   2921.00   2932.00  
6     67.38     67.37     66.11     67.10  
7     54.72     52.04     55.59     54.57  
8    421.00    431.00    421.00    390.00  
9     26.97     27.35     26.71     24.73

Data Cleaning and Preparation

In [4]:
df.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

In [283]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
60938    False
60939    False
60940    False
60941    False
60942    False
Length: 60943, dtype: bool

In [284]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB


Question 1: What is the total number and percentage of missing data in 2014 3dp

In [285]:
#What is the toal number and percetage of missing data in 2014 to 3dp

isnull=df[["Y2014"]].isnull().sum()

In [291]:
isnull


Y2014    1589
dtype: int64

The total number of missing value in 2014 is 1589

In [301]:
#Percentage is
print(" % of missing values in 2014:" , np.round(df["Y2014"].isnull().mean()*100, 3),"%")


 % of missing values in 2014: 2.607 %


Question 5: What is the total number of unique countries in the dataset?

In [274]:

df.Area.nunique()

49

Question 6- What is the total sum of Wine produced in 2015 and 2018

In [229]:

df.groupby("Item")[["Y2015", "Y2018"]].sum()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

Question 7 - Select columns Y2017 and Area. Which of these areas had the highest sum in 2017

In [276]:

df[["Y2017", "Area"]].groupby("Area").sum().sort_values(by = "Y2017")

Y2017
Area                                   
Sudan (former)                     0.00
Ethiopia PDR                       0.00
Comoros                           59.84
Seychelles                       442.34
Sao Tome and Principe          12662.63
Cabo Verde                     14650.74
Guinea-Bissau                  19102.77
Lesotho                        21267.96
Botswana                       22101.30
Djibouti                       22729.91
Gambia                         23154.18
Gabon                          27979.64
Liberia                        29342.20
Namibia                        29874.89
Central African Republic       29937.00
Congo                          41181.68
Togo                           49841.88
Mauritius                      51114.83
Eswatini                       54343.33
Sierra Leone                   55311.33
Chad                           71594.68
Rwanda                         73663.69
Zimbabwe                       75919.34
Senegal                        95681.15
Guinea                         98138.87
Burkina Faso                  101855.07
Zambia                        103223.77
Tunisia                       124167.20
Benin                         124771.22
Niger                         126707.58
Madagascar                    131197.73
Mali                          149928.33
Mauritania                    156665.46
Mozambique                    161407.98
Malawi                        181098.71
Uganda                        213950.38
Côte d'Ivoire                 224599.01
Angola                        229159.57
Cameroon                      232030.43
Sudan                         239931.92
Kenya                         264660.66
United Republic of Tanzania   322616.85
Algeria                       325644.27
Ghana                         337599.06
Morocco                       388495.36
Ethiopia                      448683.76
South Africa                  517590.54
Egypt                         866379.92
Nigeria                      1483268.23

In [236]:
df[["Y2017", "Area"]].groupby("Area").sum().sort_values(by = "Y2017").max()

Y2017    1483268.23
dtype: float64

In [240]:
df.shape    #question 11

(60943, 12)

Question 12

In [302]:
#question 12
df[["Y2017", "Area"]].groupby("Area").sum().sort_values(by = "Y2017").nsmallest(7, "Y2017")


Y2017
Area                           
Sudan (former)             0.00
Ethiopia PDR               0.00
Comoros                   59.84
Seychelles               442.34
Sao Tome and Principe  12662.63
Cabo Verde             14650.74
Guinea-Bissau          19102.77

In [11]:
df.Item

0           Population
1           Population
2          Grand Total
3          Grand Total
4          Grand Total
             ...      
60938    Miscellaneous
60939    Miscellaneous
60940    Miscellaneous
60941    Miscellaneous
60942    Miscellaneous
Name: Item, Length: 60943, dtype: object

In [12]:
(df["Item"].unique()) 

array(['Population', 'Grand Total', 'Vegetal Products', 'Animal Products',
       'Cereals - Excluding Beer', 'Wheat and products',
       'Rice and products', 'Barley and products', 'Maize and products',
       'Rye and products', 'Oats', 'Millet and products',
       'Sorghum and products', 'Cereals, Other', 'Starchy Roots',
       'Cassava and products', 'Potatoes and products', 'Sweet potatoes',
       'Yams', 'Roots, Other', 'Sugar Crops', 'Sugar cane', 'Sugar beet',
       'Sugar & Sweeteners', 'Sugar (Raw Equivalent)',
       'Sweeteners, Other', 'Honey', 'Pulses', 'Beans', 'Peas',
       'Pulses, Other and products', 'Treenuts', 'Nuts and products',
       'Oilcrops', 'Soyabeans', 'Groundnuts (Shelled Eq)',
       'Sunflower seed', 'Rape and Mustardseed', 'Cottonseed',
       'Coconuts - Incl Copra', 'Sesame seed',
       'Olives (including preserved)', 'Oilcrops, Other',
       'Vegetable Oils', 'Soyabean Oil', 'Groundnut Oil',
       'Sunflowerseed Oil', 'Rape and Mustard Oil

Question 13
What is the mean and standard deviation across the whole dataset for the year 2017 to 2 dp?

In [243]:
data = {"mean": [df["Y2017"].mean()],
        "standard_deviation": [df["Y2015"].std()]}

In [244]:
frame_mean_std = pd.DataFrame(data, index = ["Y2017"])

In [246]:
frame_mean_std.round(2)
No answer 

mean  standard_deviation
Y2017  140.92              1603.4

In [247]:
S = [["him", "sell"], [90,28,43]] #question 14
S[0][1][1]

'e'

Question 15 What year has the highest sum of import quantity

In [141]:
frame=df.groupby("Element")['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']

D:\Windows\Temp\ipykernel_4084\1145761794.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  frame=df.groupby("Element")['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']


In [155]:
frame.sum()

Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein supply quantity (g/capita/day)    11842.45    11833.56  
Residuals                                 35500.00    34864.00  
Seed                                      24870.14    25263.14  
Stock Variation                           54316.91    20577.91  
Total Population - Both sexes           1112641.00  1140605.00  
Tourist consumption                          91.00       90.00

In [248]:
frame.sum().loc["Stock Variation"]

Y2014    58749.83
Y2015    34910.99
Y2016    33140.12
Y2017    54316.91
Y2018    20577.91
Name: Stock Variation, dtype: float64

Year 2014 has the highest sum of Stock Variation

Question 16 - What is the total protein supply quantiy in Madagascar in 2015

In [279]:
protein_supply = df[(df.Area == "Madagascar")&(df.Element == "Protein supply quantity")].Y2015.sum()
protein_supply

0.0

NO ANSWER IN THE OPTIONS

Question 18
Which year has the least correlation with "Element Code"

In [280]:
df_ele_code=df[['Element Code','Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']].corr().abs()

In [281]:
df_ele_code

Element Code     Y2014     Y2015     Y2016     Y2017     Y2018
Element Code      1.000000  0.024457  0.023889  0.023444  0.024254  0.024279
Y2014             0.024457  1.000000  0.994647  0.996081  0.995230  0.994872
Y2015             0.023889  0.994647  1.000000  0.995739  0.988048  0.988208
Y2016             0.023444  0.996081  0.995739  1.000000  0.992785  0.992757
Y2017             0.024254  0.995230  0.988048  0.992785  1.000000  0.998103
Y2018             0.024279  0.994872  0.988208  0.992757  0.998103  1.000000

In [282]:
df_ele_code["Element Code"].iloc[1:].min()

0.02344399758496104

From this RESULT, the least correlation of Element code is in Y2015

Question 20
What is total number of the sum of production in 2017

In [266]:
frame_processing=df.groupby(["Element"])[['Y2017']].sum()

In [267]:
frame_processing

Y2017
Element                                           
Domestic supply quantity                2088198.10
Export Quantity                          182338.80
Fat supply quantity (g/capita/day)        10253.84
Feed                                     223705.68
Food                                    1258888.28
Food supply (kcal/capita/day)            454681.00
Food supply quantity (kg/capita/yr)       48690.04
Import Quantity                          294559.09
Losses                                   160614.00
Other uses (non-food)                     91645.97
Processing                               292836.00
Production                              2030056.89
Protein supply quantity (g/capita/day)    11842.45
Residuals                                 35500.00
Seed                                      24870.14
Stock Variation                           54316.91
Total Population - Both sexes           1112641.00
Tourist consumption                          91.00

In [278]:
frame_processing.loc["Processing"]

Y2017    292836.0
Name: Processing, dtype: float64